### ISSUE TITLE
Bring the router service to engines

### ISSUE THREAD

**Issue by buschtoens** [ember-engines/ember-engines#587]
Would be really cool, if we could bring the `RouterService` to engines as well. I propose keeping the exact same API, but adding an `*External` variant for all methods, just like we are doing with the route extension:

https://github.com/ember-engines/ember-engines/blob/ec4d1ae7a413a7e5d9e57a4e3b2e0f0d19a0afcd/addon/-private/route-ext.js#L14-L17

Implementing a proper `RouterService` for engines would also allow us to refactor the `route-ext` and also add the missing `*External` methods to `Route` and `Controller` for symmetry.

I've done a PoC service which I am using for my projects, but it would be great to have an official solution that does not rely on dirty hacks.

#### Links

- [Ember API docs for `RouterService`](https://www.emberjs.com/api/ember/3.4/classes/RouterService)
- [RFC #95: Router Service](https://github.com/emberjs/rfcs/blob/master/text/0095-router-service.md)
- [PoC implementation using some private API](https://gist.github.com/buschtoens/de2faacfc8977e39843b7e2cc2f3fd1e)

**Comment by buschtoens** [comment]
Closes #260, #391.

**Comment by villander** [comment]
@buschtoens this PoC works  making transitionToExternal & replaceWithExternal with URL?

we need pass a URL (a string that starts with a `/`).

```js
this.transitionTo('/');
this.transitionTo('/blog/post/1/comment/13');
this.transitionTo('/blog/posts?sort=title');
```

**Comment by buschtoens** [comment]
No, this intentionally not supported, as it would break the encapsulation principle of engines. You can only `transitionToExternal` routes by the name that was provided to the engine via [`externalRoutes`](https://github.com/ember-engines/ember-engines#linking-to-external-routes).

If you need to perform arbitrary navigation, yoo could [inject the host's `router` service](https://github.com/ember-engines/ember-engines#declaring-dependencies) into the engine as `host-router` / `external-router`. But I would encourage you to check whether you absolutely need to do this or if there isn't another way or abstraction you could put inbetween, because this'll also break with the encapsulation principle.

**Comment by villander** [comment]
@buschtoens sorry my comment was a bit confused, I'm saying that `transitionExternal` methods today only accept routeName but not routPath/URL, we must add support for this, as well.

**Comment by buschtoens** [comment]
@villander No worries, I understood you right the first time. ☺️

The idea behind engines is to encapsulate a part of the application from the rest. Any references / dependencies to the host app should be explicit and declarative. This is why services are explicitly injected into the engine. The same goes for routes ([`externalRoutes`](https://github.com/ember-engines/ember-engines#linking-to-external-routes)).

Allowing users to `transitionToExternal` by URL would break with that principle. The fully qualified URL can only be known by the host app. This detail should thus not be exposed to / accessible by the engine. The engine only knows about the `externalRoutes` names.

The need to transition by URL from inside an engine is an indicator for a leaky engine that is not fully encapsulated. Imagine it this way: You should be able to take that engine and drop it into any other app and be able to run it, as long as the expected external `services` and `externalRoutes` were passed into the engine.

If you still, for whatever reason, need to trigger a transition by URL from inside the engine, you can pass the host app's "real" `router` service into the engine as `external-router` and call `transitionTo` on it directly.

**Comment by rwjblue** [comment]
> The fully qualified URL can only be known by the host app. This detail should thus not be exposed to / accessible by the engine. The engine only knows about the externalRoutes names.

Completely agreed.

**Comment by villander** [comment]
yup, makes all sense @buschtoens I'll get this to work and open a pull request ASAP.

thanks!

**Comment by sumeetattree** [comment]
I think this is related?  #648

**Comment by buschtoens** [comment]
@sumeetattree Yes indeed. #648 is a duplicate of this issue / will be fixed by resolving issue. For the time being, you could use my [`EngineRouterService` PoC](https://gist.github.com/buschtoens/de2faacfc8977e39843b7e2cc2f3fd1e).

**Comment by villander** [comment]
@buschtoens I left a comment on your gist https://gist.github.com/buschtoens/de2faacfc8977e39843b7e2cc2f3fd1e#gistcomment-2960117

**Comment by buschtoens** [comment]
So we've been working with our own `EngineRouterService` for over a year now, and only really hit one issue:

The existing eco-system does not play well with `transitionToExternal`, `replaceWithExternal` and the other `*External` methods. I would propose to drop them everywhere and instead merge the internal and external namespaces, asserting that an external route cannot have the same name as an internal route.

**Comment by villander** [comment]
> The existing eco-system does not play well with transitionToExternal, replaceWithExternal and the other *External methods.

What is not working fine? can you explain more and provide some examples?

**Comment by buschtoens** [comment]
What I mean is that the whole addon eco-system uses the standard `transitionTo` and `replaceWith` method names. Addons that e.g. use the core `RouterService` have no notion of external routes and would need to be updated to support linking to external routes from inside an engine.

I only see little value in having the `*External` variants in the first place. Yes, they make it more explicit whether you are linking out of the engine at the callsite, but is this worth the incompatibility?

Instead, what if external and internal routes would be merged and would have to use shared unique names? i.e. if you pass in an external route as `foo`, you can't have an internal route named `foo`.

IMO this would remove a lot of existing maintenance burden. For instance, we could drop `<LinkToExternal>` in favor of just `<LinkTo>`, once it uses the `RouterService` under the hood.

Since this kind of a big decision, I'll branch this out into a new issue.

**Comment by villander** [comment]
transitory addon until #669 is merged - https://www.npmjs.com/package/ember-engines-router-service.

**Comment by buschtoens** [comment]
PSA for anyone subscribed to this issue: @SergeAstapov has submitted [RFC 819 "Public router service"](https://github.com/emberjs/rfcs/pull/819), which makes the `RouterService` available to engines in accordance with #691 "Merge internal and external route namespace, drop `*External` methods & `<LinkToExternal>`".
